src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                        return TryProcessJoinClause((JoinClauseSyntax)queryClause, processingState);
+                }
+
+                // OrderBy is not supported by foreach.
+                return false;
+            }
+
+            private bool TryProcessSelectClause(
+                SelectClauseSyntax selectClause,
+                QueryContinuationSyntax queryContinuation,
+                ProcessingState processingState)
+            {
+                // TODO semantic check for selectClause.Expression
+                // End of the queue
+                if (queryContinuation == null)
+                {
+                    processingState.AddStatementProcessingMethod(
+                        s => WrapWithBlock(processingState.LeafExpressionCreationMethod(selectClause.Expression)));
+                    return true;
+                }
+                else
+                {
+                    SymbolInfo symbolInfo = _semanticModel.GetSymbolInfo(queryContinuation, _cancellationToken);
+                    if (symbolInfo.Symbol == null ||
+                        symbolInfo.Symbol is IMethodSymbol methodSymbol &&
+                        methodSymbol.Parameters.Length == 1)
+                    {
+                        processingState.AddVariableName(queryContinuation.Identifier.ValueText);
+                        processingState.AddStatementProcessingMethod(
+                            s => AddToBlockTop(
+                                 SyntaxFactory.LocalDeclarationStatement(
+                                    SyntaxFactory.VariableDeclaration(
+                                        VarNameIdentifier,
+                                        SyntaxFactory.SingletonSeparatedList(
+                                            SyntaxFactory.VariableDeclarator(
+                                                queryContinuation.Identifier,
+                                                argumentList: null,
+                                                SyntaxFactory.EqualsValueClause(selectClause.Expression))))), s));
+
+                        return TryProcessQueryBody(queryContinuation.Body, processingState);
+                    }
+                }
+
+                return false;
+            }
+
+            private bool TryProcessWhereClause(WhereClauseSyntax whereClause, ProcessingState processingState)
+            {
+                if (IsValidWhereClauseSymbol(whereClause))
+                {
+                    processingState.AddStatementProcessingMethod(
+                         s => SyntaxFactory.Block(
+                             SyntaxFactory.IfStatement(
+                                 ProcessCondition((whereClause).Condition),
+                                 s)));
+
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool IsValidWhereClauseSymbol(WhereClauseSyntax whereClause)
```

Comment:
>IsValidWhereClauseSymbol [](start = 25, length = 24)

I think that the only thing this function should do, it to check return type of the predicate (bool) in addition to the method's name and number of parameters. We might also want to check that the parameter type of the predicate matches what we believe the element type is supposed to be in this part of the query, but we are not even trying to check this aspect right now, that is fine if that is intentional choice. Checking, the number of generic type parameter, return type and all other things feels wrong

Reply:
agreed to skip verification of query clauses for now

---
In reply to: [177197610](https://github.com/dotnet/roslyn/pull/25362#discussion_r177197610) [](ancestors = 177197610)